src/gax/src/error/core_error.rs

```diff
@@ -12,161 +12,664 @@
-    /// # use google_cloud_gax::error::HttpError;
-    /// use http::HeaderMap;
-    /// let error = Error::rpc(HttpError::new(404, HeaderMap::new(), None));
-    /// if let Some(e) = error.as_inner::<HttpError>() {
-    ///     assert_eq!(e.status_code(), 404);
-    /// }
-    /// ```
-    pub fn as_inner<T: std::error::Error + Send + Sync + 'static>(&self) -> Option<&T> {
-        let mut error = self.source.as_ref() as &(dyn std::error::Error);
-        loop {
-            match error.downcast_ref::<T>() {
-                Some(e) => return Some(e),
-                None => error = error.source()?,
+    /// A problem reported by the transport layer.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn http(status_code: u16, headers: HeaderMap, payload: bytes::Bytes) -> Self {
+        let kind = ErrorKind::Transport {
+            status_code: Some(status_code),
+            headers: Some(Box::new(headers)),
+            payload: Some(payload),
+            source: None,
+        };
+        Self { kind }
+    }
+
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// A problem in the transport layer without headers a full HTTP response.
+    ///
+    /// Examples include: a broken connection after the request is sent, or a
+    /// HTTP error that is *not*
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn io<T: Into<BoxError>>(source: T) -> Self {
+        Self {
+            kind: ErrorKind::Transport {
+                status_code: None,
+                headers: None,
+                payload: None,
+                source: Some(source.into()),
+            },
+        }
+    }
+
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// A problem in the transport layer without a full HTTP response.
+    ///
+    /// Examples include read or write problems, and broken connections.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn is_io(&self) -> bool {
+        matches!(
+            &self.kind,
+            ErrorKind::Transport {
+                status_code: None,
+                headers: None,
+                payload: None,
+                ..
             }
+        )
+    }
+
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// A problem reported by the transport layer.
+    #[cfg_attr(not(feature = "_internal-semver"), doc(hidden))]
+    pub fn transport<T: Into<BoxError>>(headers: HeaderMap, source: T) -> Self {
+        let kind = ErrorKind::Transport {
+            headers: Some(Box::new(headers)),
+            source: Some(source.into()),
+            status_code: None,
+            payload: None,
+        };
+        Self { kind }
+    }
+
+    /// Not part of the public API, subject to change without notice.
+    ///
+    /// A problem in the transport layer.
+    ///
+    /// Examples include I/O problems, broken connections or streams, etc.
```

Comment:
seems weird to say I/O problems when we have `is_io()` as an error class.

Fine if they are not mutually exclusive.

Reply:
PTAL.

Reply:
And I think we need a clear split between "transport" errors (include HTTP and gRPC) and I/O errors.  And if we cannot draw a clear distinction then we should keep the predicates hidden.